### PR TITLE
feat(toolbars): restructured toolbar configuration and presets

### LINKED ISSUE #508: Simplify Toolbar Configuration: Modularization and Flexibility
# 1. Introduction to the Concept of Modules

## Why Are Modules Necessary?

During the development and use of **MarkdownEditor**, there is ambiguity in terminology. The term "extension" can refer to:
- Add-ons for the WYSIWYG editor (based on **ProseMirror**),
- Add-ons for the Markdown markup editor module (based on **CodeMirror**),
- Plugins for Markdown formatting (YFM),
- Elements of the **MarkdownEditor** interface,
- Or something else entirely.

This ambiguity complicates support and development, increases code coupling, and makes scalability more difficult.

To address these issues, I propose the concept of **modules** — components that belong to the program's interface and represent its shared elements. Modules will help differentiate between internal mechanisms (ProseMirror, CodeMirror) and interface elements, providing a clearer architecture and simplifying design, implementation, and maintenance.

### Differentiating the Term "Editor":

- **MarkdownEditor** — the main program that integrates modules, extensions, and plugins. It is a unified platform for content editing.
- **WYSIWYG editor module** — handles visual text editing using ProseMirror.
- **Markdown markup editor module** — enables text editing with Markdown formatting based on CodeMirror.

### Editor Modes and Modules:

**MarkdownEditor** provides two operating modes:
1. **WYSIWYG mode**:
   Utilizes the **WYSIWYG editor module** and associated toolbar modules.
2. **Markdown markup mode**:
   Utilizes the **Markdown markup editor module** and associated toolbar modules.

### Editor Modules:

1. **WYSIWYG editor module**:
   Provides visual text editing with real-time content updates. It is based on ProseMirror and supports functional extensions via the `extensions` folder.
   
   editor_wysiwyg

2. **Markdown markup editor module**:
   Allows direct work with Markdown formatting. It is based on CodeMirror.
   
   editor_markup

### Toolbar Modules:

1. **Main toolbar in WYSIWYG mode**:
   Includes basic editing tools.
   
   toolbar_main_wysiwyg

2. **Additional toolbar in WYSIWYG mode**:
   Found in the three-dot menu and provides access to advanced tools.
   
   toolbar_additional_wysiwyg

3. **Main toolbar in Markdown markup mode**:
   Contains tools for working with Markdown formatting.
   
   tollbar_main_markup

4. **Additional toolbar in Markdown markup mode**:
   Found in the three-dot menu and offers additional Markdown-specific tools.
   
   toolbar_additional_markup

5. **Selection toolbar in WYSIWYG mode**:
   Appears when text is selected and offers contextual actions.
   
   toolbar_selection_wysiwyg

6. **Slash toolbar in WYSIWYG mode**:
   Activates upon entering the `/` character, allowing quick block selection.
   
   toolbar_slash_wysiwyg

# 2. Structure of Toolbar Buttons

A toolbar button in **MarkdownEditor** is a universal component combining an interface element with functionality for the WYSIWYG editor and Markdown markup editor modules. It should seamlessly integrate into the program's interface while executing actions tied to the **ProseMirror** or **CodeMirror** APIs.

A button can be used in any of the toolbar modules.

Given its universal nature, I propose dividing the button into three independent parts.

### Key Parts of Toolbar Buttons:

1. **Appearance**:
  - Defined within the common area of MarkdownEditor.
  - Specifies the button's visual presentation: icon, text label, and tooltip.
  - Ensures a consistent design across modes.

2. **Actions for the WYSIWYG editor module**:
  - Defined in **extensions for the WYSIWYG editor**.
  - Dictates the button's behavior in WYSIWYG mode.
  - Utilizes the ProseMirror API.

3. **Actions for the Markdown markup editor module**:
  - Defined in **Markdown markup editor configuration files**.
  - Specifies the button's behavior in Markdown markup mode.
  - Utilizes the CodeMirror API.

# 3. Presets for Toolbars

In the current version, toolbar button configurations for WYSIWYG and Markdown modes are defined separately. While this provides flexibility, it also lacks transparency.

I propose making presets the single point of entry for editor configuration.

### Objectives of Presets:

1. **Toolbar Formation**:
   Presets define the structure of toolbars, including the buttons and their order.

2. **Flexible Management**:
   To reduce the core editor's complexity, presets should be passed as external props. This allows users to create their own configurations and decouple the editor's functionality. Additionally, users can import and reuse one of the five predefined presets if needed.

3. **Support for Default Presets**:
   The editor will include five predefined presets, with three of them aligned with MarkdownIt standards:
  - **zero** — minimal button set,
  - **commonmark** — aligned with the CommonMark standard,
  - **default** — basic configuration.

### How It Works:

1. If a user passes a reserved preset name (**zero**, **commonmark**, **default**), the editor automatically applies the corresponding configuration.
2. For all other cases, a user-provided configuration object is expected.

### PR DESCRIPTION
**Pull request by makhnatkin**
Fixes #508 

1. Added new toolbars type ToolbarsPreset
2. Items are divided into three parts, with plans to place WYSIWYG elements in extension-specific folders.
3. Presets became the key point of configuration: a dictionary of elements + order, eliminating code duplication.
4. Updated presets stories, added custom presets

### CODE REVIEW

File: src/modules/toolbars/types.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+import {HotkeyProps} from '@gravity-ui/uikit';
+import {EditorState} from 'prosemirror-state';
+
+import {ActionStorage} from '../../core';
+import {CodeEditor} from '../../markup';
+import {ToolbarBaseProps, ToolbarDataType, ToolbarIconData} from '../../toolbar';
+
+// Items
+export type ToolbarItemId = string;
```
Comment (makhnatkin) [inline src/modules/toolbars/types.ts:11]: `string & {}`

File: src/modules/toolbars/types.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+export type ToolbarListId = string;
+export type ToolbarList = Record<ToolbarListId, ToolbarItemId[]>;
```
Comment (makhnatkin) [inline src/modules/toolbars/types.ts:13]: `
id: 
items:
`

File: src/modules/toolbars/types.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    title?: string | (() => string);
+    type?: ToolbarDataType;
+    doNotActivateList?: boolean;
+} & (T extends ToolbarDataType.ListButton
+    ? {
+          withArrow?: boolean;
+      }
+    : {});
+
+export interface EditorCommand<E> {
```
Comment (makhnatkin) [inline src/modules/toolbars/types.ts:29]: `EditorActions`

File: src/modules/toolbars/types.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                condition?: ((state: EditorState) => void) | 'enabled';
+            }
+          : {});
+
+export type ToolbarItemWysiwyg<T extends ToolbarDataType = ToolbarDataType.SingleButton> =
+    ToolbarItem<T, ActionStorage>;
+export type ToolbarItemMarkup<T extends ToolbarDataType = ToolbarDataType.SingleButton> =
+    ToolbarItem<T, CodeEditor>;
+
+export type ToolbarNode<T extends ToolbarDataType = ToolbarDataType.SingleButton> = {
```
Comment (makhnatkin) [inline src/modules/toolbars/types.ts:59]: `ToolbarItem`

File: src/bundle/config/markup.tsx [hunk 5]
```diff
@@ -542,209 +544,3 @@ export const mToolbarConfig: MToolbarData = [
...
-        mCodeblockItemData,
-        mCheckboxButton,
-        mTableButton,
-        mImagePopupData,
-        mHruleItemData,
-        mEmojiItemData,
-        mFilePopupData,
-        mTabsItemData,
-    ],
-};
```
Comment (d3m1d0v) [inline src/bundle/config/markup.tsx:750]: Is this a public variables? If yes, we should not remove them

File: src/bundle/config/wysiwyg.ts [hunk 6]
```diff
@@ -587,155 +590,6 @@ export const wSelectionMenuConfig: SelectionContextConfig = [
...
-                icon: icons.code,
-                withArrow: true,
-                title: i18n.bind(null, 'code'),
-                data: [wCodeItemData, wCodeBlockItemData],
-            },
-        ],
-        [wImageItemData, wFileItemData, wTableItemData, wCheckboxItemData],
-    ],
-    full: wToolbarConfig.slice(),
-};
```
Comment (d3m1d0v) [inline src/bundle/config/wysiwyg.ts:738]: same question about public variables

File: src/bundle/toolbar/utils.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    yfm,
+    full,
+};
+
+const transformItem = (
+    type: 'wysiwyg' | 'markup',
+    item?: ToolbarItem<ToolbarDataType.SingleButton | ToolbarDataType.ListButton>,
+    id = 'unknown',
+) => {
+    console.log('item', item, id);
```
Comment (d3m1d0v) [inline src/bundle/toolbar/utils.ts:30]: `console.log`

File: src/bundle/toolbar/utils.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    default: defaultPreset,
+    yfm,
+    full,
+};
+
+const transformItem = (
+    type: 'wysiwyg' | 'markup',
+    item?: ToolbarItem<ToolbarDataType.SingleButton | ToolbarDataType.ListButton>,
+    id = 'unknown',
+) => {
```
Comment (d3m1d0v) [inline src/bundle/toolbar/utils.ts:29]: plz add return type of function, to more stricter validate returned data

File: src/bundle/toolbar/utils.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+const getDefaultPresetByName = (initialPreset: EditorPreset) => {
+    const presetName = ['zero', 'commonmark', 'default', 'yfm', 'full'].includes(initialPreset)
+        ? initialPreset
+        : 'default';
+    return defaultPresets[presetName];
+};
+
+const flattenPreset = <T extends WToolbarData | MToolbarData>(config: T) => {
+    // TODO: @makhnatkin add logic for flatten
```
Comment (d3m1d0v) [inline src/bundle/toolbar/utils.ts:86]: is this a valid todo?

File: src/bundle/toolbar/utils.ts [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    return (config[0] ?? []) as unknown as T extends WToolbarData
+        ? WToolbarItemData[]
+        : MToolbarItemData[];
+};
+
+interface GetToolbarsConfigsArgs {
+    toolbarsPreset?: ToolbarsPresetOrEditorPreset;
+    props: MarkdownEditorViewProps;
```
Comment (d3m1d0v) [inline src/bundle/toolbar/utils.ts:94]: pick only necessary keys from `MarkdownEditorViewProps`

File: src/bundle/toolbar/utils.ts [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+import {
+    MToolbarData,
+    MToolbarItemData,
+    ToolbarDataType,
+    WToolbarData,
+    WToolbarItemData,
+} from '../../toolbar';
+import {MarkdownEditorViewProps} from '../MarkdownEditorView';
+
+const defaultPresets = {
```
Comment (d3m1d0v) [inline src/bundle/toolbar/utils.ts:17]: add type: `Record<EditorPreset, ToolbarsPreset>`

File: src/bundle/toolbar/utils.ts [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    );
+
+    return toolbarData as T;
+};
+
+const getDefaultPresetByName = (initialPreset: EditorPreset) => {
+    const presetName = ['zero', 'commonmark', 'default', 'yfm', 'full'].includes(initialPreset)
+        ? initialPreset
+        : 'default';
+    return defaultPresets[presetName];
```
Comment (d3m1d0v) [inline src/bundle/toolbar/utils.ts:82]:
```js
return defaultPresets[presetName] || defaultPresets.default
```

File: src/bundle/toolbar/utils.ts [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    ToolbarsPresetOrEditorPreset,
+} from '../../modules/toolbars/types';
+import {
+    MToolbarData,
+    MToolbarItemData,
+    ToolbarDataType,
+    WToolbarData,
+    WToolbarItemData,
+} from '../../toolbar';
+import {MarkdownEditorViewProps} from '../MarkdownEditorView';
```
Comment (d3m1d0v) [inline src/bundle/toolbar/utils.ts:15]: plz add `type` modifier to the imports to indicate that only types are being imported

File: src/modules/toolbars/constants.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+const listsObj = lists.reduce<Record<ListsType<typeof lists>, string>>(
+    (obj, val) => {
+        obj[val] = val;
+        return obj;
+    },
+    {} as Record<ListsType<typeof lists>, string>,
+);
+
+export const ListName: Readonly<typeof listsObj> = listsObj;
```
Comment (d3m1d0v) [inline src/modules/toolbars/constants.ts:13]: is this constant used anywhere?

File: src/modules/toolbars/constants.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+const toolbarsObj = toolbars.reduce<Record<ToolbarType<typeof toolbars>, string>>(
+    (obj, val) => {
+        obj[val] = val;
+        return obj;
+    },
+    {} as Record<ToolbarType<typeof toolbars>, string>,
+);
+
+export const ToolbarName: Readonly<typeof toolbarsObj> = toolbarsObj;
```
Comment (d3m1d0v) [inline src/modules/toolbars/constants.ts:34]: You can use a regular enum here. Why is it not used?

File: src/modules/toolbars/constants.ts (hunk 14 above)
Comment (d3m1d0v) [inline src/modules/toolbars/constants.ts:13]: same question

File: src/modules/toolbars/items.tsx [hunk 16]
```diff
@@ -0,0 +1,813 @@
...
+    type: ToolbarDataType.SingleButton,
+    icon: icons.foldingHeading,
+    title: () => i18n('folding-heading'),
+    hint: () => i18n('folding-heading__hint'),
+};
+export const toggleHeadingFoldingItemWysiwyg: ToolbarItemWysiwyg = {
+    isActive: (editor) => editor.actions.toggleHeadingFolding?.isActive() ?? false,
+    isEnable: (editor) => editor.actions.toggleHeadingFolding?.isEnable() ?? false,
+    exec: (editor) => editor.actions.toggleHeadingFolding.run(),
+    // condition: 'enabled', // FIXME: @makhnatkin
```
Comment (d3m1d0v) [inline src/modules/toolbars/items.tsx:708]: fixme

File: src/modules/toolbars/types.ts [hunk 17]
```diff
@@ -0,0 +1,87 @@
+import {RefObject} from 'react';
+
+import {HotkeyProps} from '@gravity-ui/uikit';
+import {EditorState} from 'prosemirror-state';
+
+import {ActionStorage} from '../../core';
+import {CodeEditor} from '../../markup';
+import {ToolbarBaseProps, ToolbarDataType, ToolbarIconData} from '../../toolbar';
```
Comment (d3m1d0v) [inline src/modules/toolbars/types.ts:8]: import type

File: src/bundle/MarkdownEditorView.tsx [hunk 18]
```diff
@@ -39,9 +32,22 @@ const b = cnEditorComponent;
 export type MarkdownEditorViewProps = ClassNameProps & {
     editor?: Editor;
     autofocus?: boolean;
+    toolbarsPreset?: ToolbarsPresetOrEditorPreset;
```
Comment (d3m1d0v) [inline src/bundle/MarkdownEditorView.tsx:35]: We agreed to leave only object

File: src/bundle/MarkdownEditorView.tsx [hunk 19]
```diff
@@ -73,16 +79,24 @@ export const MarkdownEditorView = React.forwardRef<HTMLDivElement, MarkdownEdito
...
+        const {
+            wysiwygToolbarConfig,
+            markupToolbarConfig,
+            wysiwygHiddenActionsConfig,
+            markupHiddenActionsConfig,
+        } = getToolbarsConfigs({
+            toolbarsPreset,
+            props,
+            preset: editor.preset,
+        });
```
Comment (d3m1d0v) [inline src/bundle/MarkdownEditorView.tsx:98]: needs to wrap to `useMemo()`

File: src/bundle/toolbar/utils.ts (hunk 9 above)
Comment (d3m1d0v) [reply src/bundle/toolbar/utils.ts:88]: \+

File: src/modules/toolbars/items.tsx [hunk 20]
```diff
@@ -0,0 +1,813 @@
...
+    withArrow: true,
+    title: i18n.bind(null, 'list'),
+};
+
+// ---- Move list ----
+export const moveListItemView: ToolbarItemView<ToolbarDataType.ListButton> = {
+    type: ToolbarDataType.ListButton,
+    icon: icons.lift,
+    withArrow: true,
+    title: 'Move list item', // TODO: @makhnatkin add i18n
```
Comment (d3m1d0v) [inline src/modules/toolbars/items.tsx:796]: todo

File: src/modules/toolbars/types.ts [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+export interface ToolbarList {
+    id: ToolbarListId;
+    items: ToolbarItemId[];
+}
+
+export type ToolbarItemView<T extends ToolbarDataType = ToolbarDataType.SingleButton> = {
+    className?: string;
+    hint?: string | (() => string);
+    hotkey?: HotkeyProps['value'];
+    type?: ToolbarDataType;
```
Comment (d3m1d0v) [inline src/modules/toolbars/types.ts:23]: add js comment that default value is  ToolbarDataType.SingleButton

File: src/modules/toolbars/types.ts [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+export type ToolbarId = string;
+export type ToolbarOrders = (ToolbarList | ToolbarItemId)[][];
+export type ToolbarsOrders = Record<ToolbarId, ToolbarOrders>;
+
+export interface ToolbarsPreset {
+    items: ToolbarsItems;
+    orders: ToolbarsOrders;
+}
+
+export type EditorPreset = 'zero' | 'commonmark' | 'default' | 'yfm' | 'full';
```
Comment (d3m1d0v) [inline src/modules/toolbars/types.ts:85]: this is copy of `MarkdownEditorPreset` from `src/bundle/types.ts`

File: src/bundle/toolbar/utils.ts (hunk 9 above)
Comment (makhnatkin) [reply src/bundle/toolbar/utils.ts:88]: It will be decided in the next PR

**Pull request merged**